### DISCUSSION TITLE
Alpine js loop inside loop

### DISCUSSION THREAD

**Discussion by shubdragon** [alpinejs/alpine#4007] [1. Help]
I  have multiple entries section and each entries section have a select element which have few option, i can achieve this using livewire easily but i don't want to send network request again for changing something because select element option not loaded in mount so it does not show on dynamically added input , so i am using alpine to archive same result

```
<div class="mb-3" x-data="{
                entries: $wire.entangle('entries'),
                ledgers: $wire.entangle('ledgers'),
                addNewRow: function() {
                    const newKey = Object.keys(this.entries).length;
                    const newRow = {
                        drCr: 1,
                        ledgerAccount: '',
                        amount: 0,
                    };
                    this.entries[newKey] = newRow;
                }
            }">

<x-bootstrap.table :headers="$headers">

              <template x-for="(entry, key) in entries" :key="key">
                    <tr>
                          <td> blahh </td>
                          <td>
                               <template x-for="(ledger, lkey) in ledgers" :key="lkey">
                                    <span  x-text="ledger.name"></span>
                                </template>
                          </td>
                   </tr>
             </template>
</x-bootstrap.table>
</div>
```

This is simplified codes i have provided which give same error the nested template is actually a option of select but i can't run nested for loop anywhere inside parent for loop. 

and the ledgers is nothing just Ledger::get()->toArray();
**without toArray() it shows  no data in alpine js dev tools 

**If I move it outside from parent loop it start to work

The Error is: Alpine Expression Error: ledger is not defined. I have alpine js dev tools installed inside which i can see ledgers is an array of objects

![Screenshot 2024-01-30 164557](https://github.com/alpinejs/alpine/assets/20332532/9cdb98fd-e1cc-479a-9e15-c5024cf67b42)

![image](https://github.com/alpinejs/alpine/assets/20332532/31ddf75f-3c06-4af0-90bd-5984e382bfb8)

**Comment by ekwoka** [comment]
You need to use an actually unique key for `:key` not the index.

this is happening when you move things and it's not longer reffering to the correct object.

**Reply by shubdragon** [reply]
I tried by making the key unique already e.g  :key="``ledger-${lkey}``" . let me try again then

**Reply by ekwoka** [reply]
That's not unique.

Because if you move the elements around in the list, different elements will have changed keys.

**Reply by shubdragon** [reply]
the error i am getting showing ledger is not defined , if i use id a key then i am still getting same issue, and the ledgers is accessible inside  the parent loop e.g` <span x-text="ledgers[0]['id']"> </span>` 

and the whole section including the parent loop is initially hidden using x-show, is that the issue?

because if i show the section initially then it is working as expected

**so the problem is when I initially hide the whole section e.g `<div x-show="selectedVoucherType" x-transition x-cloak>` if I do not use x-show to hide it initially then the problem is gone, but I need to hide it initially**

**Reply by ekwoka** [reply]
The simplest suppression is just optional chaining. `ledger?.name` will have no error though the underlying cause is still there. But this is non-destructive error

**Reply by shubdragon** [reply]
yes it will prevent the error, but actually the child loop is not working.

Three things i can do :

1. Remove x-show, so it will work as expected. (Tested)

2.  use @foreach of variable directly (Tested working)

2. should i move x-data outside of x-show and try (Not yet tried, if this the issue ?)

**Reply by ekwoka** [reply]
what version of alpine are you using?

I do actually remember issues with loops with the first child having x-data, but it was fixed I think in 3.10

**Reply by shubdragon** [reply]
it is v3.13.5, i could post more code if need but i put all relevant code except that x-show part it just a parent div above that code

**Reply by ekwoka** [reply]
And this happens on load? not like after data has changed?

What is the `ledgers` data on initial load before it does any entangling stuff?

**Reply by shubdragon** [reply]
-  initially on page it is hidden by `x-show` (**note/remember if i do not use x-show issue will no longer be issue)
- as soon as x-show true, then all entangled entries, and ledgers already available because both was initialized on livewire mount method.
- then the issue only with child x-for (if i let it load without x-show it worked).

**Comment by shubdragon** [comment]
This is some more code if need to see

Livewire mount method:
```
public function mount()
{
        $this->voucherDate = Carbon::now()->format('Y-m-d');

        $this->entries = [
            [
                'drCr' => DebitOrCreditEnum::Credit->value,
                'ledgerAccount' => '',
                'amount' => 0.00,
            ],
            [
                'drCr' => DebitOrCreditEnum::Debit->value,
                'ledgerAccount' => '',
                'amount' => 0.00,
            ],
        ];  

       $this->ledgers = Ledger::get(['id', 'name'])->toArray();     //This will be more complex query later but currently testing with this simple one and getting issue 
}
```

The View

```
....blahh with x-data="{selectedVoucherType: false}"
<form id="createAccountTransaction" wire:submit.prevent="createUpdateTransaction" method="POST" enctype="multipart/form-data" class="h-100 d-flex">
    @csrf
    <div class="row align-items-center ms-auto me-auto" x-show="!selectedVoucherType" x-transition x-cloak>
        <h1 class="display-6">{{ __('messages.selectVoucherType') }}</h1>
    </div>

    <div x-show="selectedVoucherType" x-transition x-cloak>
        {{-- some blahh div --}}
        <div class="mb-3" x-data="{
            entries: $wire.entangle('entries'),
            ledgers: $wire.entangle('ledgers'),
            addNewRow: function(drCr = 1, amount = 0.00) {
                const newKey = Object.keys(this.entries).length;                
                const newRow = {
                    drCr: drCr,
                    ledgerAccount: '',
                    amount: amount,
                };
                this.entries[newKey] = newRow;
                // ignore this this is my current solution if i use @foreach $dispatch('refreshLivewireData');
            }
        }">
            <x-bootstrap.table :headers="$headers">

                <template x-for="(entry, key) in entries" :key="`entry-${key}`">
                    <tr>                        
                        <td class="d-flex">                                                       
                            <select x-model="entry.drCr" class="form-control">
                                @foreach (DebitOrCredit::cases() as $balanceType)
                                    <option value="{{ $balanceType->value }}">{{ DebitOrCredit::shortName($balanceType) }}</option>
                                @endforeach
                            </select>
                        </td>
                        <td>
                            <select x-model="entry.ledgerAccount" x-bind:name="`entries[${key}][ledgerAccount]`" class="form-control">
                                <option value="">{{ __('messages.select') }}</option>                                
                                {{ -- @foreach ($ledgers as $ledger)
                                    <option value="{{ $ledger->id }}">{{ $ledger->name }}</option>
                                @endforeach  ignore it is my current solution --}}
                                {{-- **This is the issue part** --}}
                                <template x-for="(ledger, lkey) in ledgers" :key="`ledger-${lkey}`">
                                    <option x-bind:value="ledger?.id" x-text="ledger?.name"></option>
                                </template>
                            </select> 
                            
                        </td>
                        //blahh                        
                    </tr>
                </template>
            </x-bootstrap.table>
          {-- //blahh --}}   
    </div>
</form>
```


https://github.com/alpinejs/alpine/assets/20332532/c1b51805-44a2-4ebb-a846-08698ef7c9f3
(Ignore the UI : ))

**Reply by ekwoka** [reply]
Thanks I think I see the issue and will try to recreate. I have some in progress work that should solve this (and related bugs). Just need to design the right tests and validate it.
